Use a sensible hash for http(s) source caching

dtrudg opened the pull request:
## Description of the Pull Request (PR):

The existing caching for `http` and `https` sources uses the final part of the URL as the key for the image within the cache. This means that `http://example.com/container1/v1.sif` and `http://example.com/container2/v1.sif` result in the same cache key, and unexpected behaviour as in #4823.

We will cache using a sha256 over the URL and the date of the file that is to be
fetched, as returned by an HTTP HEAD call. If no date is available, use the current
date-time, which will effectively result in no caching.

### This fixes or addresses the following GitHub issues:

 - Fixes: #4823

#### Before submitting a PR, make sure you have done the following:

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)


Attn: @singularity-maintainers

dtrudg committed:
Use a sensible hash for http(s) source caching

Fixes: #4823

We will cache using a sha256 over the URL and the date of the file that is to be
fetched, as returned by an HTTP HEAD call. If no date is available, use the current
date-time, which will effectively result in no caching.

cclerget commented on cmd/internal/cli/actions.go:
```diff
@@ -290,11 +294,35 @@ func handleShub(imgCache *cache.Handle, u string) (string, error) {
+	imageDate := time.Now().String()
 
-	exists, err := imgCache.NetImageExists("hash", imageName)
+	req, err := http.NewRequest("HEAD", u, nil)
```
Maybe trim `http://`, `https://` from the hash as some server allow access via http and/or redirect http to https

dtrudg replied:
That's true - but equally a crazy setup could serve different things on http and http(s). That would be really stupid and unlikely, but I'd rather err on the side of re-downloading an image unless you feel strongly about this?

cclerget replied:
Not really it was just a suggestion

dtrudg committed:
Add regression test for issue4823

Check that caching behaviour is correct for http sources with same last
path part, and when dates are updated.

ikaneshiro commented on e2e/actions/regressions.go:
```diff
@@ -213,3 +218,100 @@ func (c actionTests) issue4836(t *testing.T) {
+		// Different URL - not cached
+		{
+			name:         "http_c2v2.sif_uncached",
+			urlPath:      "/c1/v2.sif",
```
Should this url be `/c2/v1.sif` instead? If the cache was keyed with just the last part of the url, as it was previously, then wouldn't cause a collision if I understand the issue correctly

dtrudg replied:
Yes - you are correct. Good spot, and :man_facepalming: from me seeing as this is what the test is meant to be for

ikaneshiro reviewed (commented):
Caching approach lgtm
